### PR TITLE
Model rollouts for training and inference

### PR DESCRIPTION
**Pull request by mishooax**
Autoregressive prediction with multi-step loss, similar to Keisler (2021).
The rollout window size is defined in the yml config file:

```yaml
# Keisler's three training rounds were:
# Round 1. ~960,000 batches @ ~0.3 seconds per batch (4-step rollout)
# Round 2. ~90,000 batches @ ~1.0 seconds per batch (8-step rollout)
# Round 3. ~70,000 batches @ ~1.5 seconds per batch (12-step rollout)
# Each batch is an N-step rollout, with batch_size=1
rollout: 4
```

The training logic looks as follows:

```python
        batch = ... # tuple (x_{t}, x_{t+dt}, x_{t+2*dt}, ... x_{t + (rollout+1)*dt})
        # start rollout
        x = batch.X[0]
        for rstep in range(self.rollout):
            y_hat = self(x)  # prediction at rollout step rstep
            y = batch.X[rstep + 1]  # target
            # y includes the auxiliary variables, so we must leave those out when computing the loss
            train_loss += self.loss(y_hat, y[..., : self.feature_dim])
            # autoregressive predictions - we re-init the "variable" part of x
            x[..., : self.feature_dim] = y_hat
```

Also: added [activation checkpointing](https://pytorch.org/docs/stable/checkpoint.html) for the GNN MLP layers. This reduces total GPU memory usage significantly, allowing us to train the following configs on a single GPU:

- (`hidden-dim: 64`, `rollout: 12`) - 2.5M model params, batch size = 1
- (`hidden-dim: 128`, `rollout: 8`) - 16M model params, batch size = 1

The "largest" (`hidden-dim: 256`, 120M param) model can't seem to fit on one GPU, even with a rollout window of length 1.

Work is on-going to scale this to multiple GPUs using DDP.

### CODE REVIEW

**Comment by mishooax** [comment]
@mchantry Hi Mat - can you take a look at this? Thanks!

File: graph_weather/data/wb_dataset.py [hunk 1]
```diff
@@ -59,22 +64,42 @@ def __init__(
 
         # pressure levels
         self.plevs = plevs
-        self.nlev = len(self.plevs) if self.plevs is not None else self._WB_PLEVS
+        self.nlev = len(self.plevs) if self.plevs is not None else constants._WB_PLEV
```
[the remark this answers is not in the document]
Comment (mchantry) [reply graph_weather/data/wb_dataset.py:65]: nlev is the number of levels.

File: graph_weather/data/wb_datamodule.py [hunk 2]
```diff
@@ -163,5 +193,91 @@ def val_dataloader(self) -> DataLoader:
...
+        )
+
+    def transfer_batch_to_device(self, batch: WeatherBenchDataBatch, device: torch.device, dataloader_idx: int = 0) -> None:
+        del dataloader_idx  # not used
+        batch.X = tuple(x.to(device) for x in batch.X)
+        batch.idx = batch.idx.to(device)
+        return batch
+
+
+class WeatherBenchTestDataModule(pl.LightningDataModule):
```
[the remark this answers is not in the document]
Comment (mishooax) [reply graph_weather/data/wb_datamodule.py:216]: Quite a lot yes, and we can improve the design. It'll have to wait for a future PR, though.

**Review by mchantry — COMMENTED** [review]
All looks sensible. Suggest waiting for current test before approving.

**Review by mchantry — APPROVED** [review]
Changes all look good.
Code now runs.
